### PR TITLE
Feature/gg 43 creating face id foundation

### PR DESCRIPTION
**Pull request by curlyjay**
Updated faceId.py program and added photoNames.txt file; The faceId.py program tests the face id functions, such as adding an employee, removing an employee, listing current employees, training face models, and recognizing faces in realtime. The photoNames.txt file is used to keep track of the arranged order of photos for each employee.

NOTE: During the OpenCV installation, while in the virtual environment ("source ~/.profile && workon cv") make sure to install the pillow package ("sudo apt-get install pillow"). 

Additionally, the FaceID code will only work while in its respective virtual environment, not under Django's. As we combine functionalities of FaceID and Django, we will have them operate under the same virtual env.

### CODE REVIEW

File: FaceID/faceID.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    trainDataset() # trains images for consistency
+
+    
+
+
+# Removes the facial recog data and database info of a specified employee
+def removeEmployee():
+
+    global photoRegister # brings in global 
+    #photoRegister = readPhotoRegister()
```
Comment (lb-cake) [inline FaceID/faceID.py]: Debugging comment can be removed.

File: FaceID/faceID.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+
+        # Define min window size to be recognized as a face
+        minW = 0.1*cam.get(3)
+        minH = 0.1*cam.get(4)
+
+        print("Press 'ESC' key to exit\n")
+        #EmpID = 0 # stores index for identified face
+        while True:
+            ret, img =cam.read()
+            #img = cv2.flip(img, -1) # Flips cam view vertically
```
Comment (lb-cake) [inline FaceID/faceID.py]: Debugging comment can be removed.

File: FaceID/faceID.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    global photoRegister # takes updated photoRegister from global variable
+
+    photoNames = open("photoNames.txt", "r") # opens txt file
+    photoRegister = photoNames.readlines() # reads lines 
+    
+    # formats read lines into 2D list
+    for i in range(len(photoRegister)):
+        person = photoRegister[i]
+        person = person.strip().split(",")
+        #print(person)
```
Comment (mraynor) [inline FaceID/faceID.py]: If we do not need to `print(person)`, we can delete this line

File: FaceID/faceID.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+                scaleFactor = 1.2,
+                minNeighbors = 5,
+                minSize = (int(minW), int(minH)),
+            )
+
+            # Only runs while a face is detected
+            for(x,y,w,h) in faces:
+
+                EmpID, confidence = recognizer.predict(gray[y:y+h,x:x+w]) # runs face profiles through recognizer
+                #print(EmpID, 100-confidence) # prints employee ID and match confidence percent to terminal
```
Comment (lb-cake) [inline FaceID/faceID.py]: Debugging code can be removed.

File: FaceID/faceID.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+                # Check if confidence is less them 100 ==> "0" is perfect match 
+                if ((confidence < 100) and (100-confidence)>25):
+                    msg = str(EmpID)
+                    confidence = "{0}%".format(round(100 - confidence)) # computes confidence level
+                    #cv2.rectangle(img, (x,y), (x+w,y+h), (0,255,0), 6) # outlines face with rectangle
```
Comment (lb-cake) [inline FaceID/faceID.py]: Debugging code can be removed.

File: FaceID/faceID.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+                    cv2.circle(img, (int(x+(w/2)),int(y+(h/2))), 108, (50,0,0), 3) # outlines face with circle
+                    cv2.putText(img, "EmpID: "+msg, (x+w+25,int(y+(h/2))), font, 1, (255,255,0), 2) # prints person's employee ID
+                    cv2.putText(img, "conf: "+str(confidence), (x+w+25,int(y+(h/2))+30), font, .75, (255,255,0), 2)  # prints confidence level
+                else:
+                    msg = "Unidentified"
+                    confidence = "{0}%".format(round(100 - confidence)) # computes confidence level
+                    #cv2.rectangle(img, (x,y), (x+w,y+h), (0,255,0), 6) # outlines face with rectangle
```
Comment (lb-cake) [inline FaceID/faceID.py]: Debugging code can be removed.

File: FaceID/faceID.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+                    cv2.circle(img, (int(x+(w/2)),int(y+(h/2))), 108, (0,0,255), 3) # outlines face with red circle
+                    #cv2.putText(img, "EmpID: "+msg, (x+w+25,int(y+(h/2))), font, 1, (255,255,255), 2) # prints person's employee ID
```
Comment (lb-cake) [inline FaceID/faceID.py]: Debugging code can be removed.

File: FaceID/faceID.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+    global photoRegister # brings 2D list from global scope
+    global frameCount
+    photoNum = 0 # an index inside of the 'person[]' list
+
+    # Manages input for new employee ID
+    flag = False
+    while flag==False:
+        empID = input("Enter your EMPLOYEE ID:  ") # get employee ID
+        if len(photoRegister)==0: 
+            flag=True
```
Comment (mraynor) [inline FaceID/faceID.py]: We should be consistent about spacing around operators throughout the code (`=`, `==`, `>` ...). I think preferred would be spaces on both sides

**Review by lb-cake — CHANGES_REQUESTED** [review]
Other then the outstanding comments and install instructions. It's looking really good.

File: FaceID/faceID.py (hunk 2 above)
Comment (mraynor) [inline FaceID/faceID.py]: Will we want to add this back later? If so, we can use `TODO`s to mark places in our code where we need to do something later (ex: `#TODO: uncomment out flip cam view code once {something} is done)`

File: FaceID/faceID.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+    return(NULL)
+
+# Placeholder function
+# Used to collect faces that are unknown
+def unknownFaces():
+    return(NULL)
+
+# Function still in testing
+# Takes in photoID to index person
+# Tdds new and removes old photos
```
Comment (mraynor) [inline FaceID/faceID.py]: Spelling of `Adds`

File: README.md [hunk 10]
```diff
@@ -117,4 +116,8 @@ $ python3 manage.py runserver
 
 **.circleci:** CircleCI directory. Used to automate testing.
 
-**venv:** Virtual environment.
\ No newline at end of file
+<<<<<<< HEAD
+**venv:** Virtual environment.
+=======
+**venv:** Virtual environment.
+>>>>>>> Implement project structure
```
Comment (mraynor) [inline README.md]: Looks like some Merge conflicts weren't completely resolved

**Comment by mraynor** [comment]
@curlyjay Would you mind cleaning up the commits a little bit. (Last week we went over rebasing/interactive rebasing) This article is really helpful: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History . If you need any help, I'll be available!

**Pull request merged**